File: chatbot-rag/src/backend/token-utils.js
/**
 * Utility functions for token counting and prompt optimization
 */

// Simple token counting approximation
// For more accurate counting, use a proper tokenizer library
function countTokens(text) {
  if (!text) return 0;
  
  // Rough approximation: 4 characters per token (average for English text)
  // This is a simplification - in production, use a proper tokenizer
  return Math.ceil(text.length / 4);
}

// Optimize prompt to reduce token usage
function optimizePrompt(prompt, maxTokens = 4000) {
  // Count tokens in the prompt
  const promptTokens = countTokens(prompt);
  
  if (promptTokens <= maxTokens) {
    return prompt;
  }
  
  // If prompt exceeds max tokens, we need to truncate it
  console.log(`Prompt exceeds token limit (${promptTokens}/${maxTokens}). Optimizing...`);
  
  // Parse the prompt to identify components
  const parts = parsePrompt(prompt);
  
  // Calculate how many tokens we need to remove
  const tokensToRemove = promptTokens - maxTokens + 200; // Add buffer
  
  // Prioritize which parts to truncate
  return truncatePromptParts(parts, tokensToRemove);
}

// Parse prompt into components
function parsePrompt(prompt) {
  // Simple parsing - in a real system this would be more sophisticated
  const lines = prompt.split('\n\n');
  const parts = [];
  
  let currentPart = { type: 'context', content: '', tokens: 0 };
  
  for (const line of lines) {
    if (line.startsWith('Document ')) {
      // Start a new document part
      if (currentPart.content) {
        currentPart.tokens = countTokens(currentPart.content);
        parts.push(currentPart);
      }
      currentPart = { type: 'document', content: line + '\n', tokens: 0 };
    } else if (line.startsWith('User question:')) {
      // User question part
      if (currentPart.content) {
        currentPart.tokens = countTokens(currentPart.content);
        parts.push(currentPart);
      }
      currentPart = { type: 'question', content: line + '\n', tokens: 0 };
    } else if (line.startsWith('Please provide')) {
      // Instructions part
      if (currentPart.content) {
        currentPart.tokens = countTokens(currentPart.content);
        parts.push(currentPart);
      }
      currentPart = { type: 'instructions', content: line, tokens: 0 };
    } else {
      // Append to current part
      currentPart.content += line + '\n\n';
    }
  }
  
  // Add the last part
  if (currentPart.content) {
    currentPart.tokens = countTokens(currentPart.content);
    parts.push(currentPart);
  }
  
  return parts;
}

// Truncate prompt parts based on priority
function truncatePromptParts(parts, tokensToRemove) {
  // Prioritize parts to truncate (documents first, then context, never question or instructions)
  const documentParts = parts.filter(p => p.type === 'document');
  
  // Sort documents by relevance (assuming later documents are less relevant)
  documentParts.sort((a, b) => {
    // Extract similarity score if available
    const getScore = (part) => {
      const match = part.content.match(/similarity: (0\.\d+)/);
      return match ? parseFloat(match[1]) : 0;
    };
    
    return getScore(b) - getScore(a);
  });
  
  let removedTokens = 0;
  let truncatedParts = [...parts];
  
  // Remove least relevant documents first
  while (removedTokens < tokensToRemove && documentParts.length > 0) {
    const leastRelevantDoc = documentParts.pop();
    removedTokens += leastRelevantDoc.tokens;
    
    // Remove this document from truncatedParts
    truncatedParts = truncatedParts.filter(p => p !== leastRelevantDoc);
  }
  
  // If we still need to remove tokens, truncate remaining documents
  if (removedTokens < tokensToRemove) {
    truncatedParts = truncatedParts.map(part => {
      if (part.type === 'document' && removedTokens < tokensToRemove) {
        // Truncate this document
        const currentTokens = part.tokens;
        const tokensToRemoveFromThisPart = Math.min(
          Math.floor(currentTokens * 0.5), // Remove up to 50% of tokens
          tokensToRemove - removedTokens
        );
        
        if (tokensToRemoveFromThisPart > 0) {
          // Truncate content
          const words = part.content.split(' ');
          const truncatedWords = words.slice(0, Math.floor(words.length * 0.5));
          const truncatedContent = truncatedWords.join(' ') + '... [truncated]';
          
          const newTokens = countTokens(truncatedContent);
          removedTokens += (currentTokens - newTokens);
          
          return {
            ...part,
            content: truncatedContent,
            tokens: newTokens
          };
        }
      }
      return part;
    });
  }
  
  // Reassemble the prompt
  return truncatedParts.map(p => p.content).join('\n\n');
}

module.exports = {
  countTokens,
  optimizePrompt
};

File: chatbot-rag/src/backend/token_utils.py
"""
Token utilities for optimizing prompts and managing token usage.
"""
import re


def estimate_tokens(text: str) -> int:
    """
    Estimate the number of tokens in a text string.
    This is a rough approximation based on character count.
    
    Args:
        text: The input text
        
    Returns:
        Estimated token count
    """
    # A very rough approximation: 1 token ≈ 4 characters for English text
    return len(text) // 4


def optimize_prompt(prompt: str, max_tokens: int = 4000) -> str:
    """
    Optimize a prompt to reduce token usage while preserving important content.
    
    Args:
        prompt: The input prompt
        max_tokens: Maximum number of tokens allowed
        
    Returns:
        Optimized prompt
    """
    estimated_tokens = estimate_tokens(prompt)
    
    # If already under the limit, return as is
    if estimated_tokens <= max_tokens:
        return prompt
    
    # Split the prompt into sections
    sections = re.split(r'\n\n+', prompt)
    
    # Identify the user question (usually at the end)
    user_question = None
    for i in range(len(sections) - 1, -1, -1):
        if sections[i].lower().startswith('user question:'):
            user_question = sections[i]
            sections.pop(i)
            break
    
    # Identify the context section
    context_start = -1
    context_end = -1
    for i, section in enumerate(sections):
        if section.startswith('Here is some relevant information'):
            context_start = i
        elif context_start >= 0 and section.strip() == '':
            context_end = i
            break
    
    # If we found the context section
    if context_start >= 0:
        context_sections = sections[context_start:context_end] if context_end > 0 else sections[context_start:]
        other_sections = sections[:context_start] + (sections[context_end:] if context_end > 0 else [])
        
        # Calculate tokens for non-context parts
        other_text = '\n\n'.join(other_sections)
        other_tokens = estimate_tokens(other_text)
        
        # Calculate available tokens for context
        available_context_tokens = max_tokens - other_tokens
        if user_question:
            available_context_tokens -= estimate_tokens(user_question)
        
        # If we need to reduce context
        if available_context_tokens < estimate_tokens('\n\n'.join(context_sections)):
            # Extract individual documents from context
            documents = []
            current_doc = []
            
            for line in '\n\n'.join(context_sections).split('\n'):
                if line.startswith('Document '):
                    if current_doc:
                        documents.append('\n'.join(current_doc))
                    current_doc = [line]
                else:
                    current_doc.append(line)
            
            if current_doc:
                documents.append('\n'.join(current_doc))
            
            # Sort documents by relevance (assuming they're already sorted)
            # Keep adding documents until we hit the token limit
            optimized_context = []
            current_tokens = 0
            
            for doc in documents:
                doc_tokens = estimate_tokens(doc)
                if current_tokens + doc_tokens <= available_context_tokens:
                    optimized_context.append(doc)
                    current_tokens += doc_tokens
                else:
                    break
            
            # Rebuild the prompt
            result = []
            if context_start > 0:
                result.extend(sections[:context_start])
            
            if optimized_context:
                result.append('Here is some relevant information that might help answer the question:')
                result.extend(optimized_context)
            
            if context_end > 0:
                result.extend(sections[context_end:])
            
            if user_question:
                result.append(user_question)
            
            return '\n\n'.join(result)
    
    # Fallback: simple truncation while preserving the user question
    if user_question:
        # Calculate available tokens for the rest of the prompt
        available_tokens = max_tokens - estimate_tokens(user_question) - 100  # Buffer
        
        # Truncate the rest of the prompt
        rest_of_prompt = '\n\n'.join(sections)
        truncated_prompt = rest_of_prompt[:available_tokens * 4]  # Convert back to characters
        
        return truncated_prompt + '\n\n' + user_question
    
    # Last resort: just truncate
    return prompt[:max_tokens * 4]  # Convert tokens to approximate characters
